README.md
# image-js

[![NPM version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]
[![Test coverage][codecov-image]][codecov-url]
[![npm download][download-image]][download-url]

Advanced image processing and manipulation in JavaScript.

image-js is a full-featured library that can deal with simple image processing
(color leveling, grey image, mask, resize, rotation, etc.) as well as advanced
processing on scientific images (Region of interest (ROI), Hull curve, minimal
boundary rectangle (MBR), particle size and orientation, cell imaging, etc.).

## [API Documentation](https://image-js.github.io/image-js/)

## Installation

`$ npm install image-js`

## Features

### Supported image formats

The following formats can be loaded by image-js:

- PNG (8 or 16 bits, color or greyscale, with or without alpha, palette 1 - 8 bits)
- JPEG
- TIFF (8 or 16 bits, greyscale)

The following formats can be saved by image-js:

- PNG (8 or 16 bits)
- JPEG
- BMP (black and white)

### Native support for various bit depths and image kinds

image-js was developed to be used in scientific applications where we often
have to work on images that have more that 8 bits per channel.  
Unlike many other libraries, if a 16-bit greyscale PNG is decoded, the resulting
image has only one 16-bit channel and no pixel information is lost.

image-js can work with images that have 1 (binary), 8, 16 or 32 bits per channel.  
It can accept an arbitrary amount of color channels (usually 1 or 3) and can
handle an additional alpha component.

### Basic image manipulation

image-js can be used to do simple image manipulations such as:

- Resize
- Crop
- Rotate
- Convert to greyscale
- Invert colors
- Gaussian blur
- Extract individual channels (red, green or blue)
- And more...

### Statistics

image-js implements a number of functions to get statistics about an image:

- Histogram
- Max, min, median value
- And more ...

### Advanced features for computer vision

- Image thresholding (otsu, triangle, ...)
- Regions of interest
- Convolution with custom kernel
- Sobel filter
- Morphological transformations (open, close, erode, ...)

## An example using npm and node

Install the library:

```console
npm i image-js
```

An example of code manipulating the image 'cat.jpg' (you need to create it).

```js
const { Image } = require('image-js');

execute().catch(console.error);

async function execute() {
  let image = await Image.load('cat.jpg');
  let grey = image
    .grey() // convert the image to greyscale.
    .resize({ width: 200 }) // resize the image, forcing a width of 200 pixels. The height is computed automatically to preserve the aspect ratio.
    .rotate(30); // rotate the image clockwise by 30 degrees.
  return grey.save('cat.png');
}
```

```console
node index.js
```

A greyscale image will be saved in the same folder.

## Examples in the browser

### Load an image and convert it to grey

```html
<html>

<head>
    <script src="https://www.lactame.com/lib/image-js/0.21.2/image.min.js"></script>
</head>

<body>
    <img id="color" src="https://www.lactame.com/github/image-js/image-js/3073b80c7d626196cb669f9d617f491a8338ca66/test/img/taxi/original.jpeg" />
    <img id='result' />

    <script>
        async function process() {
            let image = await IJS.Image.load(document.getElementById('color').src);

            let grey=image.grey();

            document.getElementById('result').src = grey.toDataURL();
        }
        process();
    </script>

</body>

</html>
```

<img src="./test/img/taxi/original.jpeg" width="300" /><img src="./test/img/taxi/grey.png" width="300" />

[Try it](https://www.w3schools.com/code/tryit.asp?filename=FVD6UC9S7RSV)

### Create a mask

<img src="./test/img/taxi/original.jpeg" width="300" /><img src="./test/img/taxi/mask.png" width="300" />

[Try it](https://www.w3schools.com/code/tryit.asp?filename=FVD6TP7ICRZM)

### Paint a mask

<img src="./test/img/taxi/original.jpeg" width="300" /><img src="./test/img/taxi/orange.png" width="300" />

[Try it](https://www.w3schools.com/code/tryit.asp?filename=FVD6T1Y63MUH)

### Filter a mask using Region Of Interests (ROIs)

Image-js has a powerful Region of Interests Manager that allows to create ROIs from different sources. The ROIs can then be filtered, manipulated and finally painted to an RGBA image.

<img src="./test/img/taxi/original.jpeg" width="300" /><img src="./test/img/taxi/filteredOrange.png" width="300" />

[Try it](https://www.w3schools.com/code/tryit.asp?filename=FVD6SE2F4MWE)

When extracting a mask from a ROI you have many options (`contour`, `box`, `filled`, `center`, `hull` or `normal`). Here it looks better to use the `filled` ROI.

<img src="./test/img/taxi/original.jpeg" width="300" /><img src="./test/img/taxi/filteredFilledOrange.png" width="300" />

[Try it](https://www.w3schools.com/code/tryit.asp?filename=FVD6QQISMLZ2)

### Advanced analysis of SEM / TEM images

This library is able to deal with complex analysis involving images of cell or SEM / TEM. It will deal correctly with 16 bits grey scale images (TIFF or PNG) commonly found in scientific results.

In this example we will annotate an SEM / TEM image by coloring each particle and show the surface of them.

<img src="./test/img/semtem/semtem.jpg" width="400" />
<img src="./test/img/semtem/result.png" />

We also display a table containing a summary of all the identified particles.

[Try it](https://www.w3schools.com/code/tryit.asp?filename=FVCJLR0VNK33)

## Development

Contributions to code or documentation are welcome! Here are a few tips on how to
setup a development environment for image-js.

### Canvas

The `canvas` native addon library is required for all tests to pass. You can
follow the instructions to install it on your OS [here](https://github.com/Automattic/node-canvas#installation).

## License

[MIT](./LICENSE)

[npm-image]: https://img.shields.io/npm/v/image-js.svg?style=flat-square
[npm-url]: https://www.npmjs.com/package/image-js
[travis-image]: https://img.shields.io/travis/image-js/image-js/master.svg?style=flat-square
[travis-url]: https://travis-ci.org/image-js/image-js
[codecov-image]: https://img.shields.io/codecov/c/github/image-js/image-js.svg?style=flat-square
[codecov-url]: https://codecov.io/gh/image-js/image-js
[download-image]: https://img.shields.io/npm/dm/image-js.svg?style=flat-square
[download-url]: https://www.npmjs.com/package/image-js


History.md
<a name="0.21.4"></a>
## [0.21.4](https://github.com/image-js/core/compare/v0.21.3...v0.21.4) (2018-09-17)



<a name="0.21.3"></a>
## [0.21.3](https://github.com/image-js/core/compare/v0.21.2...v0.21.3) (2018-09-17)



<a name="0.21.2"></a>
## [0.21.2](https://github.com/image-js/core/compare/v0.21.1...v0.21.2) (2018-09-16)



<a name="0.21.1"></a>
## [0.21.1](https://github.com/image-js/core/compare/v0.21.0...v0.21.1) (2018-09-08)


### Bug Fixes

* pin canvas to alpha.12 ([1982ffb](https://github.com/image-js/core/commit/1982ffb))
* **build:** correct babel 7 config ([da0c8a9](https://github.com/image-js/core/commit/da0c8a9))
* **package:** update blob-util to version 2.0.2 ([#390](https://github.com/image-js/core/issues/390)) ([72780ce](https://github.com/image-js/core/commit/72780ce)), closes [#388](https://github.com/image-js/core/issues/388)


### Features

* **load:** add support palette PNG ([f6042f1](https://github.com/image-js/core/commit/f6042f1))



<a name="0.21.0"></a>
# [0.21.0](https://github.com/image-js/core/compare/v0.20.1...v0.21.0) (2018-04-26)


### Features

* add toBuffer method ([f2fdfc7](https://github.com/image-js/core/commit/f2fdfc7))



<a name="0.20.1"></a>
## [0.20.1](https://github.com/image-js/core/compare/v0.20.0...v0.20.1) (2018-04-13)


### Bug Fixes

* **save:** fix jpeg saving ([c4d90d0](https://github.com/image-js/core/commit/c4d90d0))



<a name="0.20.0"></a>
# [0.20.0](https://github.com/image-js/core/compare/v0.19.1...v0.20.0) (2018-03-15)


### Features

* insert image into another one ([f690a30](https://github.com/image-js/core/commit/f690a30))



<a name="0.19.1"></a>
## [0.19.1](https://github.com/image-js/core/compare/v0.19.0...v0.19.1) (2018-02-16)



<a name="0.19.0"></a>
# [0.19.0](https://github.com/image-js/core/compare/v0.18.1...v0.19.0) (2018-02-16)


### Performance Improvements

* optimize dilate and erode for "only-1" kernels with grey images ([2119414](https://github.com/image-js/core/commit/2119414))
* optimize erode and dilate for "only-1" kernels ([#374](https://github.com/image-js/core/issues/374)) ([6023f70](https://github.com/image-js/core/commit/6023f70))



<a name="0.18.1"></a>
## [0.18.1](https://github.com/image-js/core/compare/v0.18.0...v0.18.1) (2018-02-06)


### Bug Fixes

* allow again to export png from 1 or 32 bit images ([838f5e0](https://github.com/image-js/core/commit/838f5e0))



<a name="0.18.0"></a>
# [0.18.0](https://github.com/image-js/core/compare/v0.17.0...v0.18.0) (2018-02-06)



<a name="0.17.0"></a>
# [0.17.0](https://github.com/image-js/core/compare/v0.16.1...v0.17.0) (2018-02-01)


* infer file format from filename when saving ([9a30c34](https://github.com/image-js/core/commit/9a30c34))


### Bug Fixes

* fix gaussian blur. kernels should sum to 1 ([ea70851](https://github.com/image-js/core/commit/ea70851))
* pass bitDepth option in gradientFilter ([0b60bce](https://github.com/image-js/core/commit/0b60bce))


### Features

* add abs filter ([671f601](https://github.com/image-js/core/commit/671f601))
* direction option in sobelFilter ([8b9034e](https://github.com/image-js/core/commit/8b9034e))
* iteration parameter on erode ([4fd915a](https://github.com/image-js/core/commit/4fd915a))
* support conversion of 32-bit images to rgba data ([5790815](https://github.com/image-js/core/commit/5790815))


### BREAKING CHANGES

* not specifying a file format will throw instead of saving to png
* sobelFilter no longer allows custom filters. Use gradientFilter instead



<a name="0.16.1"></a>
## [0.16.1](https://github.com/image-js/core/compare/v0.16.0...v0.16.1) (2018-01-24)


### Features

* add getThreshold ([d4fdf9f](https://github.com/image-js/core/commit/d4fdf9f))



<a name="0.16.0"></a>
# [0.16.0](https://github.com/image-js/core/compare/v0.15.0...v0.16.0) (2018-01-17)


### Bug Fixes

* force version of rollup-plugin-node-resolve ([80b59bd](https://github.com/image-js/core/commit/80b59bd))
* **package:** does not support node 6 ([a1660bd](https://github.com/image-js/core/commit/a1660bd))


### Features

* decode JPEG in JS ([8d92060](https://github.com/image-js/core/commit/8d92060))



<a name="0.15.0"></a>
# [0.15.0](https://github.com/image-js/core/compare/v0.14.4...v0.15.0) (2017-12-11)



<a name="0.14.4"></a>
## [0.14.4](https://github.com/image-js/core/compare/v0.14.3...v0.14.4) (2017-08-16)


### Bug Fixes

* fix problem caused by inversion between white and black ([8eaf922](https://github.com/image-js/core/commit/8eaf922))



<a name="0.14.3"></a>
## [0.14.3](https://github.com/image-js/core/compare/v0.14.2...v0.14.3) (2017-08-15)


### Bug Fixes

* errors in the call of the morphological transforms ([b99d938](https://github.com/image-js/core/commit/b99d938))



<a name="0.14.2"></a>
## [0.14.2](https://github.com/image-js/core/compare/v0.14.1...v0.14.2) (2017-08-15)



<a name="0.14.1"></a>
## [0.14.1](https://github.com/image-js/core/compare/v0.14.0...v0.14.1) (2017-08-15)


### Bug Fixes

* errors in some cases with erode and dilate ([717d173](https://github.com/image-js/core/commit/717d173))
* errors in the definition of the dimensions of the submatrix ([e86e6a0](https://github.com/image-js/core/commit/e86e6a0))



<a name="0.14.0"></a>
# [0.14.0](https://github.com/image-js/core/compare/v0.13.0...v0.14.0) (2017-08-15)


### Bug Fixes

* add missing files ([3dc9491](https://github.com/image-js/core/commit/3dc9491))
* add missing files ([51d45ff](https://github.com/image-js/core/commit/51d45ff))
* change the arguments (replaced by one argument 'options') ([8f8ebc9](https://github.com/image-js/core/commit/8f8ebc9))
* check image and kernel ([7ea982c](https://github.com/image-js/core/commit/7ea982c))
* correction of the name of the errors ([afd79ba](https://github.com/image-js/core/commit/afd79ba))
* modify names of the tests ([c41e611](https://github.com/image-js/core/commit/c41e611))


### Features


index.d.ts
// Type definitions for image-js 0.18
// Project: https://github.com/image-js/image-js
// Definitions by: Michaël Zasso <https://github.com/targos>

// Global variable exposed by UMD bundle
export as namespace IJS;

declare class Image {
    width: number;
    height: number;
    data: DataArray;
    size: number;
    components: number;
    alpha: BinaryValue;
    bitDepth: BitDepth;
    maxValue: number;
    colorModel: ColorModel;
    channels: number;
    meta: object;
    
    constructor(width: number, height: number, data: ArrayLike<number>, options?: ImageConstructorOptions);
    constructor(width: number, height: number, options?: ImageConstructorOptions);
    constructor(options?: ImageConstructorOptions)

    static isImage(object: any): boolean;
    static fromCanvas(canvas: HTMLCanvasElement): Image;
    static createFrom(other: Image, options: ImageConstructorOptions);
    static load(image: (string|ArrayBuffer|Uint8Array), options?: RequestInit)

    getRoiManager(): RoiManager;
    clone(): Image;
    
    // valueMethods
    getValueXY(x: number, y: number, channel: number): number;
    setValueXY(x: number, y: number, channel: number, value: number): this;
    getValue(index: number, channel: number): number;
    setValue(index: number, channel: number, value: number): this;
    getPixelXY(x: number, y: number): Array<number>;
    setPixelXY(x: number, y: number, value: Array<number>): this;
    getPixel(index: number): Array<number>;
    setPixel(index: number, value: Array<number>): this;

    // bitMethods
    getBit(index: number): BinaryValue;
    setBit(index: number): void;
    clearBit(index: number): void;
    toggleBit(index: number): void;
    getBitXY(x: number, y: number): BinaryValue;
    setBitXY(x: number, y: number): void;
    clearBitXY(x: number, y: number): void;
    toggleBitXY(x: number, y: number): void;

    // exportMethods
    save(path: string, options?: SaveOptions): Promise<void>;
    toDataURL(type?: string, options?: SaveOptions): string;
    toBase64(type?: string, options?: SaveOptions): string;
    toBlob(type?: string, quality?: number): Blob;
    getCanvas(): HTMLCanvasElement;

    checkProcessable(processName: string, options: object): void;
    getRGBAData(options?: {clamped?: boolean}): (Uint8Array|Uint8ClampedArray);

    // extend
    invert(options?: OutOrInplace): Image;
    abs(options?: OutOrInplace): Image;
    level(options?: {channels?: SelectedChannels, min?: number, max?: number}): this;
    // add
    // subtract
    // subtractImage
    // multiply
    // divide
    // hypotenuse
    // background
    // flipX
    // flipY

    blurFilter(options?: {radius?: number}): Image;
    medianFilter(options?: {radius?: number, border?: BorderHandling, channels?: SelectedChannels}): Image;
    gaussianFilter(options?: GaussianFilterOptions): Image;
    gradientFilter(options?: GradientFilterOptions): Image;
    sobelFilter(options?: GradientOptions): Image;
    scharrFilter(options?: GradientOptions): Image;

    dilate(options?: MorphologicalOptions): Image;
    erode(options?: MorphologicalOptions): Image;
    open(options?: MorphologicalOptions): Image;
    close(options?: MorphologicalOptions): Image;
    topHat(options?: MorphologicalOptions): Image;
    blackHat(options?: MorphologicalOptions): Image;
    morphologicalGradient(options?: MorphologicalOptions): Image;

    // warpingFourPoints
    crop(options?: CropOptions): Image;
    // cropAlpha
    resize(options?: ResizeOptions): Image;
    // hsv
    // hsl
    // cmyk
    // rgba8
    grey(options?: GreyOptions): Image;
    mask(options?: MaskOptions): Image;
    // pad
    // colorDepth
    // setBorder
    rotate(angle: number, options?: RotateOptions): Image;
    rotateLeft(): Image;
    rotateRight(): Image;

    // getRow
    // getColumn
    // getMatrix
    // setMatrix
    // getPixelsArray
    // getIntersection
    // getClosestCommonParent
    // getThreshold

    // split
    // getChannel
    // combineChannels
    // setChannel
    // getSimilarity
    // getPixelsGrid
    // getBestMatch

    // cannyEdge
    convolution(kernel: Kernel, options?: ConvolutionOptions): Image;
    // extract
    // floodFill
    // paintLabels
    // paintMasks
    // paintPoints
    // paintPolyline
    // paintPolylines
    // paintPolygon
    // paintPolygons

    // countAlphaPixels
    // monotoneChainConvexHull
    // minimalBoundingRectangle
    // getHistogram
    // getHistograms
    // getColorHistogram
    // getMin
    // getMax
    // getSum
    // getMoment
    // getLocalMaxima
    // getMedian
    // getMean
    // getPoints
    // getRelativePosition
}

declare class RoiManager {}

interface ImageConstructorOptions {
    width?: number,
    height?: number,
    data?: ArrayLike<number>,
    kind?: ImageKind,
    bitDepth?: BitDepth,
    components?: number,
    alpha?: (0|1),
    colorModel?: ColorModel
}

interface SaveOptions {
    format?: string,
    useCanvas?: boolean,
    encoder?: object
}

interface OutOrInplace {
    inPlace?: boolean,
    out?: Image
}

interface MorphologicalOptions {
    kernel?: BinaryKernel,
    iterations?: number
}

interface GaussianFilterOptions {
    radius?: number,
    sigma?: number,
    channels?: SelectedChannels,
    border?: BorderHandling,
    algorithm?: ConvolutionAlgorithm
}

interface ConvolutionOptions {
    channels?: SelectedChannels,
    bitDepth?: BitDepth,
    normalize?: boolean,
    divisor?: number,
    border?: BorderHandling,
    algorithm?: ConvolutionAlgorithm
}

interface GradientOptions {
    direction?: GradientDirection,
    border?: BorderHandling,
    channels?: SelectedChannels,
    bitDepth?: BitDepth
}

interface GradientFilterOptions extends GradientOptions {
    kernelX?: Kernel,
    kernelY?: Kernel
}

declare enum ImageKind {
    BINARY = 'BINARY',
    GREY = 'GREY',
    GREYA = 'GREYA',
    RGB = 'RGB',
    RGBA = 'RGBA',
    CMYK = 'CMYK',
    CMYKA = 'CMYKA',
}

declare enum BitDepth {
    BINARY = 1,
    UINT8 = 8,
    UINT16 = 16,
    FLOAT32 = 32
}

declare enum ColorModel {
    GREY = 'GREY',
    RGB = 'RGB',
    HSL = 'HSL',
    HSV = 'HSV',
    CMYK = 'CMYK'
}

declare enum BorderHandling {
    COPY = 'copy'
}

declare enum ConvolutionAlgorithm {
    AUTO = 'auto',
    DIRECT = 'direct',
    FFT = 'fft',
    SEPARABLE = 'separable'
}

interface CropOptions {
    x?: number;
    y?: number;
    width?: number;
    height?: number;
}

interface ResizeOptions {
    width?: number;
    height?: number;
    factor?: number;
    interpolation?: InterpolationAlgorithm;
    preserveAspectRatio?: boolean;
}

interface GreyOptions {
    algorithm?: GreyAlgorithm|GreyAlgorithmCallback;
    keepAlpha?: boolean;
    mergeAlpha?: boolean;
    out?: Image;
}

declare enum GreyAlgorithm {
    LUMA709 = 'luma709',
    LUMA601 = 'luma601',
    MAXIMUM = 'maximum',
    MINIMUM = 'minimum',
    AVERAGE = 'average',
    MINMAX = 'minmax',
    RED = 'red',
    GREEN = 'green',
    BLUE = 'blue',
    CYAN = 'cyan',
    MAGENTA = 'magenta',
    YELLOW = 'yellow',
    BLACK = 'black',
    HUE ='hue',
    SATURATION = 'saturation',
    LIGHTNESS = 'lightness'
}

type GreyAlgorithmCallback = (red: number, green: number, blue: number) => number;

interface MaskOptions {
    algorithm?: ThresholdAlgorithm|'threshold';
    threshold?: number;
    useAlpha?: boolean;
    invert?: boolean;
}

declare enum ThresholdAlgorithm {
    HUANG = 'huang',
    INTERMODES = 'intermodes',
    ISODATA = 'isodata',
    LI = 'li',
    MAX_ENTROPY = 'maxentropy',
    MEAN = 'mean',
    MIN_ERROR = 'minerror',
    MOMENTS = 'moments',
    OTSU = 'otsu',
    PERCENTILE = 'percentile',
    RENYI_ENTROPY = 'renyientropy',
    SHANBHAG = 'shanbhag',
    TRIANGLE = 'triangle',
    YEN = 'yen'
}

interface RotateOptions {
    interpolation?: InterpolationAlgorithm
}

declare enum GradientDirection {
    WIDTH = 'x',
    HEIGHT = 'y',
    BOTH = 'xy'
}

declare enum InterpolationAlgorithm {
    NEAREST_NEIGHBOR = 'nearestNeighbor',
    BILINEAR = 'bilinear'
}

type DataArray = Uint8Array | Uint16Array | Float32Array;
type BinaryValue = 0 | 1;
type SelectedChannels = number | string | Array<number> | Array<string>;
type BinaryKernel = Array<Array<BinaryValue>>;
type Kernel = Array<Array<number>>;

export { Image }
export default Image;
